crates/matrix-sdk/src/send_queue/mod.rs

```diff
@@ -726,18 +778,25 @@ impl RoomSendQueue {
                     let media_source = if room.latest_encryption_state().await?.is_encrypted() {
                         trace!("upload will be encrypted (encrypted room)");
                         let mut cursor = std::io::Cursor::new(data);
-                        let encrypted_file = room
-                            .client()
+                        let client = room.client();
+                        let mut req = client
                             .upload_encrypted_file(&mut cursor)
-                            .with_request_config(RequestConfig::short_retry())
-                            .await?;
+                            .with_request_config(RequestConfig::short_retry());
+                        if let Some(watcher) = progress_watcher {
+                            req = req.with_send_progress_observable(watcher);
+                        };
+                        let encrypted_file = req.await?;
                         MediaSource::Encrypted(Box::new(encrypted_file))
                     } else {
                         trace!("upload will be in clear text (room without encryption)");
                         let request_config = RequestConfig::short_retry()
                             .timeout(Media::reasonable_upload_timeout(&data));
-                        let res =
-                            room.client().media().upload(&mime, data, Some(request_config)).await?;
+                        let mut req =
+                            room.client().media().upload(&mime, data, Some(request_config));
+                        if let Some(watcher) = progress_watcher {
+                            req = req.with_send_progress_observable(watcher);
```

Comment:
If I understand correctly, this will emit the following updates, if I'm sending a file with a thumbnail:

- first progress updates for the thumbnail, going from 0 to 100%
- then progress updates for the file itself, going from 0 to 100%

Is that right? I think observers would likely be more observed in the overall progress from 0 to 100%, so it would be nice if we could aggregate those into a single stream, for observers (especially for galleries which may contain many medias and thumbnails).

Reply:
I think this goes together with your [other comment](https://github.com/matrix-org/matrix-rust-sdk/pull/5008/files#r2108440747). I actually cannot think of a reason why a client would be interested in displaying the thumbnail and file upload progress separately. So aggregating them would probably be best.

I think a difficulty with this is determining the overall upload size ahead of time. The way the code is currently structured, we only receive the size of the file upload once the file has been encrypted and actually started uploading. To aggregate it, we would have to know the size already when starting the thumbnail upload, however.

Reply:
Maybe we could update `UploadEncryptedFile` so that it can either take a reader for the plain data (like today) or a vector of the encrypted data? Then we could encrypt both thumbnail and media ahead of time, to compute the overall upload size, store it on the send queue requests and use it inside `RoomSendQueue::sending_task` to aggregate the progress per thumbnail / media pair. That would mingle the concerns a little of course.

The only other option I can think of is to do some heuristics and translate the bytes sent for the encrypted attachments into bytes of the unencrypted attachments. That would mean the SDK wouldn't report the actually uploaded bytes though which doesn't feel great either.

🤔

Reply:
Oh, yes good point that about the memory limitations when encrypting everything upfront. 

I think the percentage should probably suffice in most cases. Maybe we could just map the internal `TransmissionProgress` to a new struct that contains only the current percentage then.

Reply:
I've implemented this by introducing

```
pub struct RelativeTransmissionProgress {
    /// The completion percentage as a number between 0 and 1.
    pub percentage: f32,
}
```

which worked relatively ok apart from having to add three further fields on the request kinds to track the different file sizes.

I've realized too late that a floating point percentage might introduce subtle precision issues. Maybe it's better to track the progress as a pair of `current` / `total` integers, possibly with an `fn percentage() -> f32`?

The other shortcoming is that it's not (easily) possible anymore to track _just_ the thumbnail upload progress (or completion). That is probably not a huge use case though.

Reply:
@bnjbvr do you have any thoughts on how to proceed here?

Reply:
Resolving this since `AbstractProgress` has been implemented in the latest version of the PR.